crates/re_chunk/src/chunk.rs

```diff
@@ -0,0 +1,478 @@
+use std::collections::BTreeMap;
+
+use arrow2::array::Array as ArrowArray;
+
+use re_log_types::{EntityPath, ResolvedTimeRange, RowId, TimeInt, TimePoint, Timeline};
+use re_types_core::{ComponentName, SerializationError};
+
+// ---
+
+/// Errors that can occur when creating/manipulating a [`Chunk`]s.
+#[derive(thiserror::Error, Debug)]
+pub enum ChunkError {
+    #[error("Detected malformed Chunk: {reason}")]
+    Malformed { reason: String },
+
+    #[error(transparent)]
+    Serialization(#[from] SerializationError),
+
+    #[error("Chunks cannot be empty")]
+    Empty,
+}
+
+pub type ChunkResult<T> = Result<T, ChunkError>;
+
+// ---
+
+/// Unique identifier for a [`Chunk`], using a [`re_tuid::Tuid`].
+pub type ChunkId = re_tuid::Tuid;
+
+/// Dense arrow-based storage of N rows of multi-component multi-temporal data for a specific entity.
+///
+/// This is our core datastructure for logging, storing, querying and transporting data around.
+///
+/// The chunk as a whole is always ascendingly sorted by [`RowId`] before it gets manipulated in any way.
+/// Its time columns might or might not be ascendingly sorted, depending on how the data was logged.
+///
+/// This is the in-memory representation of a chunk, optimized for efficient manipulation of the
+/// data within.
+#[derive(Debug, Clone)]
+pub struct Chunk {
+    pub(crate) id: ChunkId,
+    pub(crate) entity_path: EntityPath,
+
+    /// Is the chunk as a whole sorted by [`RowId`]?
+    pub(crate) is_sorted: bool,
+
+    /// The respective [`RowId`]s for each row of data.
+    pub(crate) row_ids: Vec<RowId>,
+
+    /// The time columns.
+    ///
+    /// Each column must be the same length as `row_ids`.
+    ///
+    /// Empty if this is a static chunk.
+    pub(crate) timelines: BTreeMap<Timeline, ChunkTimeline>,
+
+    /// A sparse `ListArray` for each component.
```

Comment:
To my knowledge arrow doesn't have a spec for "sparse" listarray.

Do you mean nullable listarray?

Reply:
Also, worth thinking about. Arrow now supports a ListView: https://arrow.apache.org/docs/format/Columnar.html#listview-layout

This could give us a mechanism to shuffle just the offsets in cases where we don't want to pay the full cost of rearranging the child buffer.

Reply:
> To my knowledge arrow doesn't have a spec for "sparse" listarray.
> 
> Do you mean nullable listarray?

I just find the "official" wording extremely confusing: what's a nullable listarray exactly? a listarray that can be null? a listarray that can contain null values? both?